### ISSUE TITLE
Question: How to timeout a cdk deployment to ecs service when task fail to start

### ISSUE THREAD

**Issue by mouyigang** [awslabs/aws-cdk#1498]
From what I observed, if the ecs fargate service(tasks) fail to start it will re-try and I never saw it actually timeout.

Is there anything we can do from CDK to timeout a deployment?

**Comment by rix0rrr** [comment]
It will timeout eventually, but it will take a while (an hour or so).

That's a good question if we can set a timeout. I actually don't know the answer to that.

cc @SoManyHs ?

**Comment by digitalsanctum** [comment]
I ran into this today as well. My fix was to kill CDK which was hung and go into the aws console to "Cancel update stack". From there the stack was rolled back to it's original state.

Seems like in these cases you'd want to fail fast(er) with a timeout override and do the above steps automatically to get back to a known state.

**Comment by rix0rrr** [comment]
That is a good point. You can use the console to interrupt the deployment. You don't even have to kill CDK to do it either, CDK will show the rollback starting and exit with an error appropriately at the end if you do.

**Comment by mouyigang** [comment]
Another issue I found related to this was: after cancel the deploy(and delete the stack in cloudformation) it didn't remove the log groups(created by the stack with specified static names) which will fail in the next cdk deploy.

I had to manually remove those log groups.

**Comment by rix0rrr** [comment]
That is actually on purpose, to always retain logs, they might be important.

There is a parameter to control that, but I suppose we can improve the default on whether a static name was specified or not

**Comment by digitalsanctum** [comment]
> Another issue I found related to this was: after cancel the deploy(and delete the stack in cloudformation) it didn't remove the log groups(created by the stack with specified static names) which will fail in the next cdk deploy.
> 
> I had to manually remove those log groups.

Yes, I ran into this as well. IMO, the defaults should be such that destroy -> deploy should be idempotent.

**Comment by rix0rrr** [comment]
I don't think there's any way to control the per-resource timeout in CloudFormation. We can control the stack-wide timeout instead though. Should be a toolkit feature.

**Comment by mipearson** [comment]
@rix0rrr Note that stack-wide timeouts seem to be limited to creation rather than update :(

**Comment by zxkane** [comment]
Looks like there is already creation timeout of nested stack. Is there any way to specify timeout of updating nested stack?

**Comment by ivawzh** [comment]
> We can control the stack-wide timeout instead though. Should be a toolkit feature.

It will be an awesome feature ☝️ ❤️ , and hopefully it will help avoid having to set up a complex manual check for ECS deployment crash loop like this  https://aws.amazon.com/blogs/compute/automating-rollback-of-failed-amazon-ecs-deployments/

On the other hand, I can confirm that currently, ECS EC2 service deployment takes 3 hours to decide that it failed, when the error comes from application layer within docker container.

```log
 3/6 | 6:33:37 AM | UPDATE_IN_PROGRESS   | AWS::ECS::Service        | xxx
 ...
 5/6 | 9:35:07 AM | UPDATE_FAILED        | AWS::ECS::Service        | xxx  Service arn:aws:ecs:xxx:xxx:service/xxx did not stabilize.
```

**Comment by acro5piano** [comment]
The current workaround is updating the service's desired count to 0.

**Comment by misterjoshua** [comment]
Would it be acceptable if the CDK CLI trapped a `WINCH` signal to signal the sending of a `cloudformation:CancelUpdateStack`? If we did this, then the user could run CDK like this: `timeout -sWINCH 15m cdk deploy ...`

**Comment by adambiggs** [comment]
Just ran into this issue.

Currently, the only workaround requires manual intervention which really goes against the whole philosophy of CDK Pipelines...

**Comment by pharindoko** [comment]
Horrible ... they should never have created cdk as wrapper for cloudformation.

**Comment by achan-godaddy** [comment]
We're having this problem as well, the workaround does work but what does one do in production if the service count can't be dropped to 0?

**Comment by rofr** [comment]
> We're having this problem as well, the workaround does work but what does one do in production if the service count can't be dropped to 0?

Cancel the CloudFormation update in the console or using the CLI, fix the problem and then redeploy.

**Comment by evgenyneu** [comment]
> You can use the console to interrupt the deployment.
> Cancel the CloudFormation update in the console or using the CLI

Sorry guys what do you mean exactly by "cancel from console"? What is "console" and how do you cancel? I'm deploying from a Terminal with command `cdk deploy`. Do you mean that the deploy will be cancelled with I press Ctrl-C?

**Comment by jakedmarcotte** [comment]
> > You can use the console to interrupt the deployment.
> > Cancel the CloudFormation update in the console or using the CLI
> 
> Sorry guys what do you mean exactly by "cancel from console"? What is "console" and how do you cancel? I'm deploying from a Terminal with command `cdk deploy`. Do you mean that the deploy will be cancelled with I press Ctrl-C?

@evgenyneu Console meaning the [AWS Management Console](https://aws.amazon.com/console/). You should be able to cancel from Stack Actions -> Cancel Update Stack

**Comment by madeline-k** [comment]
Related: https://github.com/aws/aws-cdk/issues/3781
